### DISCUSSION TITLE
Flickering in Terminal UI

### DISCUSSION THREAD

**Discussion by sluongng** [bazelbuild/bazel#23507] [Ideas]
This is probably a minor nit, but a new user might notice the flickering in the progress UI by quite a bit. There seem to be 2 main problems with the current UI:

1. The action progress flickers quite a lot when the build is on-going.

2. The cursor jumped all over the place during build, assuming because redraw process left it there.

Most of the relevant code is in [UiEventHandler.java](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/runtime/UiEventHandler.java). I am not quite sure just yet how to fix/improve this 🤔

**Comment by meisterT** [comment]
What terminal / OS are you using? I don't see any flickering (or perhaps I am not looking at the right thing).

**Reply by sluongng** [reply]
I am using Ghostty terminal on MacOS with a layer of Tmux on top.
I also see the same problem when using Kitty terminal to open the same Tmux session.
Opening a vertical Tmux split during the build showed no flickering, so that should also exclude problems originating from Tmux.

Most terminal CLI I use renders just fine.
Newer ones created in Rust are working with a very high frame rate.

Here is an asciinema I recorded just now building Bazel.git https://asciinema.org/a/OR9wg9ktndKtc2URPHTtAUe9A

If you download it and replay it with `asciinema play` frame by frame, you would be able to see instances like this:

- Frame 0:

![image](https://github.com/user-attachments/assets/cad4a12f-bc26-44d8-b9d1-c7be45692a2c)

- Frame 1:

![image](https://github.com/user-attachments/assets/260c3295-adc8-4cb2-bba8-62ffea4a0b0c)

- Frame 2:

![image](https://github.com/user-attachments/assets/61b0fae8-1566-4f56-ae68-43f0f0db405e)

- Frame 3:

![image](https://github.com/user-attachments/assets/981cfbca-0a20-4633-ad0d-d39bc488d042)

This also explains why I see the cursor jump to different places.

And if you speed up the playback, you can see the flickering intensify, matching what I experienced on my end.
